{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Script:  Automatically counts alleles in GT_variation data after filtering\"\"\"\n",
    "\n",
    "#  Used modules\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading main files\n",
    "\n",
    "##  Exones data\n",
    "exones = pd.read_excel(\"Exones_variance_pdfilter.xlsx\")\n",
    "\n",
    "##  Genes data\n",
    "genes  = pd.read_excel(\"Genes_variance_pdfilter.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Used columns - same for exones & genes\n",
    "\n",
    "CS_columns = [i for i in range(1,64)]\n",
    "CN_columns = [i for i in range(64,97) if i != 85] ## 85 skipped - duplicated in data\n",
    "\n",
    "#  Summary columns name\n",
    "\n",
    "## named in order - 1/1, 0/1, 0/0, 0/n (n - positive integer), n/n\n",
    "## numbers - related to alleles in vcf\n",
    "\n",
    "columns_case = [\"CS_AA\", \"CS_AR\", \"CS_RR\", \"CS_RO\", \"CS_OO\"]\n",
    "columns_control = [\"CN_AA\",\"CN_AR\",\"CN_RR\", \"CN_RO\", \"CN_OO\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Alleles counting exone data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy main dataset - against possible fragmentations\n",
    "\n",
    "exones_copy = exones.copy()\n",
    "\n",
    "\n",
    "\n",
    "#  Division data by : → extraction only GT for sample\n",
    "\n",
    "## CS\n",
    "\n",
    "for i in CS_columns:\n",
    "    exones_copy[f\"GT_{i}\"] = exones_copy[i].astype(str).str.split(\":\").str[0]\n",
    "\n",
    "## CN\n",
    "    \n",
    "for i in CN_columns:\n",
    "    exones_copy[f\"GT_{i}\"] = exones_copy[i].astype(str).str.split(\":\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CS_RR\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:8: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CS_RO\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CS_OO\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:14: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CN_AA\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CN_AR\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:16: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CN_RR\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:17: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CN_RO\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:18: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CN_OO\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:50: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CS_A\"] = exones_copy[\"CS_AA\"] * 2 + exones_copy[\"CS_AR\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:51: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CS_R\"] = exones_copy[\"CS_RR\"] * 2 + exones_copy[\"CS_AR\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:52: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CN_A\"] = exones_copy[\"CN_AA\"] * 2 + exones_copy[\"CN_AR\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:53: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CN_R\"] = exones_copy[\"CN_RR\"] * 2 + exones_copy[\"CN_AR\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:58: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CS_O\"] = exones_copy[\"CS_OO\"] * 2 + exones_copy[\"CS_RO\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:60: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"CN_O\"] = exones_copy[\"CN_OO\"] * 2 + exones_copy[\"CN_RO\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\593134813.py:67: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  exones_copy[\"Total\"] = np.where(exones_copy[\"CS_AA\"] + exones_copy[\"CS_AR\"] + exones_copy[\"CS_RR\"] + exones_copy[\"CN_AA\"] + exones_copy[\"CN_AR\"] + exones_copy[\"CN_RR\"] == 95, 1, 0)\n"
     ]
    }
   ],
   "source": [
    "#  creating aummary columns\n",
    "\n",
    "## CS\n",
    "\n",
    "exones_copy[\"CS_AA\"] = 0\n",
    "exones_copy[\"CS_AR\"] = 0\n",
    "exones_copy[\"CS_RR\"] = 0\n",
    "exones_copy[\"CS_RO\"] = 0\n",
    "exones_copy[\"CS_OO\"] = 0\n",
    "\n",
    "\n",
    "## CN\n",
    "\n",
    "exones_copy[\"CN_AA\"] = 0\n",
    "exones_copy[\"CN_AR\"] = 0\n",
    "exones_copy[\"CN_RR\"] = 0\n",
    "exones_copy[\"CN_RO\"] = 0\n",
    "exones_copy[\"CN_OO\"] = 0\n",
    "\n",
    "# adding specific genotypes\n",
    "    \n",
    "## CS\n",
    "    \n",
    "for i in CS_columns:\n",
    "    ###  Counts 1/1, 0/1, 0/1\n",
    "    exones_copy[\"CS_AA\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^1/1').astype(int)\n",
    "    exones_copy[\"CS_AR\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/1').astype(int)\n",
    "    exones_copy[\"CS_RR\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/0').astype(int)\n",
    "\n",
    "    ###  Counts 0/n, n/n - n = positive integer 1 to 19\n",
    "    exones_copy[\"CS_RO\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/[1-9]|1[0-9]').astype(int)\n",
    "    exones_copy[\"CS_OO\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^[1-9]|1[0-9]/[1-9]|1[0-9]').astype(int)\n",
    "\n",
    "## CN\n",
    "\n",
    "for i in CN_columns:\n",
    "    ###  Counts 1/1, 0/1, 0/1\n",
    "    exones_copy[\"CN_AA\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^1/1').astype(int)\n",
    "    exones_copy[\"CN_AR\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/1').astype(int)\n",
    "    exones_copy[\"CN_RR\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/0').astype(int)\n",
    "\n",
    "    ###  Counts 0/n, n/n - n = positive integer 1 to 19\n",
    "    exones_copy[\"CN_RO\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/[1-9]|1[0-9]').astype(int)\n",
    "    exones_copy[\"CN_OO\"] += exones_copy[f\"GT_{i}\"].astype(str).str.count(r'^[1-9]|1[0-9]/[1-9]|1[0-9]').astype(int)\n",
    "\n",
    "#  Creating alleles columns\n",
    "    \n",
    "##  aleles 0 and 1\n",
    "\n",
    "exones_copy[\"CS_A\"] = exones_copy[\"CS_AA\"] * 2 + exones_copy[\"CS_AR\"]\n",
    "exones_copy[\"CS_R\"] = exones_copy[\"CS_RR\"] * 2 + exones_copy[\"CS_AR\"]\n",
    "exones_copy[\"CN_A\"] = exones_copy[\"CN_AA\"] * 2 + exones_copy[\"CN_AR\"]\n",
    "exones_copy[\"CN_R\"] = exones_copy[\"CN_RR\"] * 2 + exones_copy[\"CN_AR\"]\n",
    "\n",
    "##  aleles 0 and others\n",
    "\n",
    "exones_copy[\"CS_R\"] = exones_copy[\"CS_RR\"] * 2 + exones_copy[\"CS_RO\"]\n",
    "exones_copy[\"CS_O\"] = exones_copy[\"CS_OO\"] * 2 + exones_copy[\"CS_RO\"]\n",
    "exones_copy[\"CN_R\"] = exones_copy[\"CN_RR\"] * 2 + exones_copy[\"CN_RO\"]\n",
    "exones_copy[\"CN_O\"] = exones_copy[\"CN_OO\"] * 2 + exones_copy[\"CN_RO\"]\n",
    "\n",
    "#  Checking if sum of columns equal 95\n",
    "    \n",
    "##  1 == total is 95\n",
    "##  0 == total not 95\n",
    "    \n",
    "exones_copy[\"Total\"] = np.where(exones_copy[\"CS_AA\"] + exones_copy[\"CS_AR\"] + exones_copy[\"CS_RR\"] + exones_copy[\"CN_AA\"] + exones_copy[\"CN_AR\"] + exones_copy[\"CN_RR\"] == 95, 1, 0)\n",
    "\n",
    "\n",
    "#  Creating two datasets - multi (multi alt), singl (only one alt)\n",
    "\n",
    "exones_multi = exones_copy.loc[exones_copy[\"Total\"] == 0]\n",
    "exones_single = exones_copy.loc[exones_copy[\"Total\"] == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Export final files\n",
    "\n",
    "## Multi\n",
    "\n",
    "exones_multi.to_excel(\"Exones_variance_pdfilter_allelesdone_multi.xlsx\", index=False)\n",
    "\n",
    "## Single\n",
    "\n",
    "exones_single.to_excel(\"Exones_varace_pdfilter_allelesdone_single.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Alleles counting genes data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Copy main dataset agains fragmentation\n",
    "\n",
    "genes_copy = genes.copy()\n",
    "\n",
    "\n",
    "#  Division data by : → extraction only GT for sample\n",
    "\n",
    "## CS\n",
    "\n",
    "for i in CS_columns:\n",
    "    genes_copy[f\"GT_{i}\"] = genes_copy[i].astype(str).str.split(\":\").str[0]\n",
    "\n",
    "## CN\n",
    "    \n",
    "for i in CN_columns:\n",
    "    genes_copy[f\"GT_{i}\"] = genes_copy[i].astype(str).str.split(\":\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CS_RR\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:8: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CS_RO\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CS_OO\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:14: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CN_AA\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CN_AR\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:16: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CN_RR\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:17: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CN_RO\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:18: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CN_OO\"] = 0\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:50: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CS_A\"] = genes_copy[\"CS_AA\"] * 2 + genes_copy[\"CS_AR\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:51: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CS_R\"] = genes_copy[\"CS_RR\"] * 2 + genes_copy[\"CS_AR\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:52: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CN_A\"] = genes_copy[\"CN_AA\"] * 2 + genes_copy[\"CN_AR\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:53: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CN_R\"] = genes_copy[\"CN_RR\"] * 2 + genes_copy[\"CN_AR\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:58: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CS_O\"] = genes_copy[\"CS_OO\"] * 2 + genes_copy[\"CS_RO\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:60: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"CN_O\"] = genes_copy[\"CN_OO\"] * 2 + genes_copy[\"CN_RO\"]\n",
      "C:\\Users\\chumi\\AppData\\Local\\Temp\\ipykernel_12080\\3602881067.py:68: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  genes_copy[\"Total\"] = np.where(genes_copy[\"CS_AA\"] + genes_copy[\"CS_AR\"] + genes_copy[\"CS_RR\"] + genes_copy[\"CN_AA\"] + genes_copy[\"CN_AR\"] + genes_copy[\"CN_RR\"] == 95, 1, 0)\n"
     ]
    }
   ],
   "source": [
    "#  creating aummary columns\n",
    "\n",
    "## CS\n",
    "\n",
    "genes_copy[\"CS_AA\"] = 0\n",
    "genes_copy[\"CS_AR\"] = 0\n",
    "genes_copy[\"CS_RR\"] = 0\n",
    "genes_copy[\"CS_RO\"] = 0\n",
    "genes_copy[\"CS_OO\"] = 0\n",
    "\n",
    "\n",
    "## CN\n",
    "\n",
    "genes_copy[\"CN_AA\"] = 0\n",
    "genes_copy[\"CN_AR\"] = 0\n",
    "genes_copy[\"CN_RR\"] = 0\n",
    "genes_copy[\"CN_RO\"] = 0\n",
    "genes_copy[\"CN_OO\"] = 0\n",
    "\n",
    "# adding specific genotypes\n",
    "    \n",
    "## CS\n",
    "    \n",
    "for i in CS_columns:\n",
    "    ###  Counts 1/1, 0/1, 0/1\n",
    "    genes_copy[\"CS_AA\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^1/1').astype(int)\n",
    "    genes_copy[\"CS_AR\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/1').astype(int)\n",
    "    genes_copy[\"CS_RR\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/0').astype(int)\n",
    "\n",
    "    ###  Counts 0/n, n/n - n = positive integer 1 to 19\n",
    "    genes_copy[\"CS_RO\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/[1-9]|1[0-9]').astype(int)\n",
    "    genes_copy[\"CS_OO\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^[1-9]|1[0-9]/[1-9]|1[0-9]').astype(int)\n",
    "\n",
    "## CN\n",
    "\n",
    "for i in CN_columns:\n",
    "    ###  Counts 1/1, 0/1, 0/1\n",
    "    genes_copy[\"CN_AA\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^1/1').astype(int)\n",
    "    genes_copy[\"CN_AR\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/1').astype(int)\n",
    "    genes_copy[\"CN_RR\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/0').astype(int)\n",
    "\n",
    "    ###  Counts 0/n, n/n - n = positive integer 1 to 19\n",
    "    genes_copy[\"CN_RO\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^0/[1-9]|1[0-9]').astype(int)\n",
    "    genes_copy[\"CN_OO\"] += genes_copy[f\"GT_{i}\"].astype(str).str.count(r'^[1-9]|1[0-9]/[1-9]|1[0-9]').astype(int)\n",
    "\n",
    "#  Creating alleles columns\n",
    "\n",
    "##  aleles 0 and 1\n",
    "\n",
    "genes_copy[\"CS_A\"] = genes_copy[\"CS_AA\"] * 2 + genes_copy[\"CS_AR\"]\n",
    "genes_copy[\"CS_R\"] = genes_copy[\"CS_RR\"] * 2 + genes_copy[\"CS_AR\"]\n",
    "genes_copy[\"CN_A\"] = genes_copy[\"CN_AA\"] * 2 + genes_copy[\"CN_AR\"]\n",
    "genes_copy[\"CN_R\"] = genes_copy[\"CN_RR\"] * 2 + genes_copy[\"CN_AR\"]\n",
    "\n",
    "##  aleles 0 and others\n",
    "\n",
    "genes_copy[\"CS_R\"] = genes_copy[\"CS_RR\"] * 2 + genes_copy[\"CS_RO\"]\n",
    "genes_copy[\"CS_O\"] = genes_copy[\"CS_OO\"] * 2 + genes_copy[\"CS_RO\"]\n",
    "genes_copy[\"CN_R\"] = genes_copy[\"CN_RR\"] * 2 + genes_copy[\"CN_RO\"]\n",
    "genes_copy[\"CN_O\"] = genes_copy[\"CN_OO\"] * 2 + genes_copy[\"CN_RO\"]\n",
    "\n",
    "\n",
    "#  Checking if sum of columns equal 95\n",
    "    \n",
    "##  1 == total is 95\n",
    "##  0 == total not 95\n",
    "    \n",
    "genes_copy[\"Total\"] = np.where(genes_copy[\"CS_AA\"] + genes_copy[\"CS_AR\"] + genes_copy[\"CS_RR\"] + genes_copy[\"CN_AA\"] + genes_copy[\"CN_AR\"] + genes_copy[\"CN_RR\"] == 95, 1, 0)\n",
    "\n",
    "\n",
    "#  Creating two datasets - multi (multi alt), singl (only one alt)\n",
    "\n",
    "genes_multi = genes_copy.loc[genes_copy[\"Total\"] == 0]\n",
    "genes_single = genes_copy.loc[genes_copy[\"Total\"] == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Export final files\n",
    "\n",
    "## Multi\n",
    "\n",
    "genes_multi.to_excel(\"Genes_variance_pdfilter_allelesdone_multi.xlsx\", index=False)\n",
    "\n",
    "## Single\n",
    "\n",
    "genes_single.to_excel(\"Genes_varace_pdfilter_allelesdone_single.xlsx\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
