{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Needed modules\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading files\n",
    "\n",
    "\n",
    "#  Fst statistics files\n",
    "\n",
    "Fst_exones = pd.read_excel(\"FstExones_Case_vs_Control.xlsx\").fillna(\"NaN\")\n",
    "Fst_genes = pd.read_excel(\"FstGenes_Case_vs_Control.xlsx\").fillna(\"NaN\")\n",
    "\n",
    "\n",
    "#  Files - no empty genotypes in all subject; filtered by Cadd_phred > 15\n",
    "\n",
    "main_exones = pd.read_excel(\"Exones_variance_pdfilter_allelesdone_single.xlsx\")\n",
    "main_genes = pd.read_excel(\"Genes_variance_pdfilter_allelesdone_single.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Copy against fragmentation\n",
    "\n",
    "#  Main files\n",
    "\n",
    "main_exones_copy = main_exones.copy()\n",
    "main_genes_copy = main_genes.copy()\n",
    "\n",
    "#  Fst files\n",
    "\n",
    "Fst_exones_copy = Fst_exones.copy()\n",
    "Fst_genes_copy = Fst_genes.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Creating double CN_alleles\n",
    "\n",
    "##  Exones\n",
    "\n",
    "main_exones_copy[\"CN_2A\"] = main_exones_copy[\"CN_A\"] * 2\n",
    "main_exones_copy[\"CN_2R\"] = main_exones_copy[\"CN_R\"] * 2\n",
    "\n",
    "##  Genes\n",
    "\n",
    "main_genes_copy[\"CN_2A\"] = main_genes_copy[\"CN_A\"] * 2\n",
    "main_genes_copy[\"CN_2R\"] = main_genes_copy[\"CN_R\"] * 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting Fst data according chromosom and position from used files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Creating Fst column for Fst scores data\n",
    "\n",
    "#  Combining CHROM|POS - Combined columns\n",
    "\n",
    "##  Exones\n",
    "\n",
    "Fst_exones_copy[\"Combined_Fst\"] = Fst_exones_copy[\"CHROM\"].astype(str) + \"_\" + Fst_exones_copy[\"POS\"].astype(str)\n",
    "main_exones_copy[\"Combined_main_exones\"] = main_exones_copy[\"CHROM\"].astype(str) + \"_\" + main_exones_copy[\"POS\"].astype(str)\n",
    "\n",
    "##  Genes\n",
    "\n",
    "Fst_genes_copy[\"Combined_Fst\"] = Fst_genes_copy[\"CHROM\"].astype(str) + \"_\" + Fst_genes_copy[\"POS\"].astype(str)\n",
    "main_genes_copy[\"Combined_main_genes\"] = main_genes_copy[\"CHROM\"].astype(str) + \"_\" + main_genes_copy[\"POS\"].astype(str)\n",
    "\n",
    "#  Merging dataframes → setting similar combined data\n",
    "\n",
    "##  Exones\n",
    "\n",
    "merged_exones = Fst_exones_copy.merge(main_exones_copy, left_on='Combined_Fst', right_on='Combined_main_exones', how='left')\n",
    "\n",
    "##  Genes\n",
    "\n",
    "merged_genes = Fst_genes_copy.merge(main_genes_copy,left_on='Combined_Fst', right_on='Combined_main_genes', how='left')\n",
    "\n",
    "\n",
    "#  Filtering only usefull columns from merged dataframes\n",
    "\n",
    "##  Exones\n",
    "\n",
    "merged_exones_position = merged_exones.loc[:,[\"CHROM_x\", \"POS_x\", \"WEIR_AND_COCKERHAM_FST\", \"Combined_Fst\", \"CHROM_y\", \"POS_y\", \"Combined_main_exones\"]].reset_index(drop=True)\n",
    "\n",
    "##  Genes\n",
    "\n",
    "merged_genes_position = merged_genes.loc[:,[\"CHROM_x\", \"POS_x\", \"WEIR_AND_COCKERHAM_FST\", \"Combined_Fst\", \"CHROM_y\", \"POS_y\", \"Combined_main_genes\"]].reset_index(drop=True)\n",
    "\n",
    "\n",
    "#  Setting Fst columns\n",
    "\n",
    "##  Exones\n",
    "\n",
    "merged_exones_filtered = merged_exones_position[merged_exones_position[\"Combined_main_exones\"].notna()].reset_index(drop=True)\n",
    "\n",
    "merged_exones_filtered_dict = dict(zip(merged_exones_filtered[\"Combined_main_exones\"],merged_exones_filtered[\"WEIR_AND_COCKERHAM_FST\"]))\n",
    "\n",
    "main_exones_copy[\"Fst\"] = main_exones_copy[\"Combined_main_exones\"].map(merged_exones_filtered_dict).fillna(\"NaN\")\n",
    "\n",
    "##  Genes\n",
    "\n",
    "merged_genes_filtered = merged_genes_position[merged_genes_position[\"Combined_main_genes\"].notna()].reset_index(drop=True)\n",
    "\n",
    "merged_genes_filtered_dict = dict(zip(merged_genes_filtered[\"Combined_main_genes\"],merged_genes_filtered[\"WEIR_AND_COCKERHAM_FST\"]))\n",
    "\n",
    "main_genes_copy[\"Fst\"] = main_genes_copy[\"Combined_main_genes\"].map(merged_genes_filtered_dict).fillna(\"NaN\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Application of Fishers exact test (scipy.stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FET\n",
    "\n",
    "##  Exones\n",
    "\n",
    "FET1_exones = main_exones_copy.apply(lambda x: scipy.stats.fisher_exact([[x.CS_A,x.CS_R],[x.CN_A, x.CN_R]]), axis=1)\n",
    "FET2_exones = main_exones_copy.apply(lambda x: scipy.stats.fisher_exact([[x.CS_A,x.CS_R],[x.CN_2A, x.CN_2R]]), axis=1)\n",
    "\n",
    "##  Genes\n",
    "\n",
    "FET1_genes = main_genes_copy.apply(lambda x: scipy.stats.fisher_exact([[x.CS_A,x.CS_R],[x.CN_A, x.CN_R]]), axis=1)\n",
    "FET2_genes = main_genes_copy.apply(lambda x: scipy.stats.fisher_exact([[x.CS_A,x.CS_R],[x.CN_2A, x.CN_2R]]), axis=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Application Odds ratio\n",
    "\n",
    "- scipy.stats.contingency.odds_ratio\n",
    "- output:\n",
    "- confidence interval - only for more details\n",
    "- odds ratio - main result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  OR\n",
    "\n",
    "##  Exones\n",
    "\n",
    "### OR1\n",
    "\n",
    "OR1_exones = main_exones_copy.apply(lambda x: scipy.stats.contingency.odds_ratio([[x.CS_A,x.CS_R],[x.CN_A, x.CN_R]], kind='sample'), axis=1)  #  Applying OR\n",
    "\n",
    "OR1_exones_results = OR1_exones.apply(lambda res: res.statistic) #  Getting result\n",
    "\n",
    "\n",
    "###  OR2\n",
    "\n",
    "OR2_exones = main_exones_copy.apply(lambda x: scipy.stats.contingency.odds_ratio([[x.CS_A,x.CS_R],[x.CN_2A, x.CN_2R]], kind='sample'), axis=1)  #  Applying OR\n",
    "\n",
    "OR2_exones_results = OR2_exones.apply(lambda res: res.statistic) #  Getting result\n",
    "\n",
    "#  Genes\n",
    "\n",
    "#  OR1\n",
    "\n",
    "OR1_genes = main_genes_copy.apply(lambda x: scipy.stats.contingency.odds_ratio([[x.CS_A, x.CS_R], [x.CN_A, x.CN_R]], kind='sample'), axis=1)  #  Applying OR\n",
    "\n",
    "OR1_genes_results = OR1_genes.apply(lambda res: res.statistic)  #  Getting result\n",
    "\n",
    "\n",
    "#  OR2\n",
    "\n",
    "OR2_genes = main_genes_copy.apply(lambda x: scipy.stats.contingency.odds_ratio([[x.CS_A, x.CS_R], [x.CN_2A, x.CN_2R]], kind='sample'), axis=1)  #  Applying OR\n",
    "\n",
    "OR2_genes_results = OR2_genes.apply(lambda res: res.statistic)  #  Getting result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\chumi\\Documents\\Programming_space\\.venv\\Lib\\site-packages\\scipy\\stats\\_odds_ratio.py:155: RuntimeWarning: divide by zero encountered in log\n",
      "  log_or = np.log(oddsratio)\n",
      "c:\\Users\\chumi\\Documents\\Programming_space\\.venv\\Lib\\site-packages\\scipy\\stats\\_odds_ratio.py:156: RuntimeWarning: divide by zero encountered in divide\n",
      "  se = np.sqrt((1/table).sum())\n",
      "c:\\Users\\chumi\\Documents\\Programming_space\\.venv\\Lib\\site-packages\\scipy\\stats\\_odds_ratio.py:169: RuntimeWarning: invalid value encountered in scalar add\n",
      "  loghigh = log_or + z*se\n",
      "c:\\Users\\chumi\\Documents\\Programming_space\\.venv\\Lib\\site-packages\\scipy\\stats\\_odds_ratio.py:168: RuntimeWarning: invalid value encountered in scalar subtract\n",
      "  loglow = log_or - z*se\n"
     ]
    }
   ],
   "source": [
    "#  Adding confidence interval\n",
    "\n",
    "#  Exones\n",
    "\n",
    "OR1_exones_CI = OR1_exones.apply(lambda res: res.confidence_interval(confidence_level=0.95))\n",
    "\n",
    "OR2_exones_CI = OR2_exones.apply(lambda res: res.confidence_interval(confidence_level=0.95))\n",
    "\n",
    "#  Genes\n",
    "\n",
    "OR1_genes_CI = OR1_genes.apply(lambda res: res.confidence_interval(confidence_level=0.95))\n",
    "\n",
    "OR2_genes_CI = OR2_genes.apply(lambda res: res.confidence_interval(confidence_level=0.95))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Extraction FET p-values, Odds ratio and Confidence interval (low, high)\n",
    "\n",
    "#  Exones\n",
    "\n",
    "#  P-values\n",
    "\n",
    "main_exones_copy[\"FET1\"] = FET1_exones.apply(lambda x: x[1]).fillna(\"NaN\")\n",
    "main_exones_copy[\"FET2\"] = FET2_exones.apply(lambda x: x[1]).fillna(\"NaN\")\n",
    "\n",
    "#  OR\n",
    "\n",
    "main_exones_copy[\"OR1\"] = OR1_exones_results.fillna(\"NaN\")\n",
    "main_exones_copy[\"OR2\"] = OR2_exones_results.fillna(\"NaN\")\n",
    "\n",
    "\n",
    "#  Confidence interval\n",
    "\n",
    "#  Regular CN counts group - interval - <low,high>\n",
    "\n",
    "main_exones_copy[\"OR1_low\"] = OR1_exones_CI.apply(lambda res: res.low).fillna(\"NaN\")\n",
    "main_exones_copy[\"OR1_high\"] = OR1_exones_CI.apply(lambda res: res.high).fillna(\"NaN\")\n",
    "\n",
    "\n",
    "#  Double CN counts group - interval - <low, high>\n",
    "\n",
    "main_exones_copy[\"OR2_low\"] = OR2_exones_CI.apply(lambda res: res.low).fillna(\"NaN\")\n",
    "main_exones_copy[\"OR2_high\"] = OR2_exones_CI.apply(lambda res: res.high).fillna(\"NaN\")\n",
    "\n",
    "\n",
    "#  Genes\n",
    "\n",
    "#  P-values\n",
    "\n",
    "main_genes_copy[\"FET1\"] = FET1_genes.apply(lambda x: x[1]).fillna(\"NaN\")\n",
    "main_genes_copy[\"FET2\"] = FET2_genes.apply(lambda x: x[1]).fillna(\"NaN\")\n",
    "\n",
    "#  OR\n",
    "\n",
    "main_genes_copy[\"OR1\"] = OR1_genes_results.fillna(\"NaN\")\n",
    "main_genes_copy[\"OR2\"] = OR2_genes_results.fillna(\"NaN\")\n",
    "\n",
    "#  Confidence interval\n",
    "\n",
    "#  Regular CN counts group - interval - <low,high>\n",
    "\n",
    "main_genes_copy[\"OR1_low\"] = OR1_genes_CI.apply(lambda res: res.low).fillna(\"NaN\")\n",
    "main_genes_copy[\"OR1_high\"] = OR1_genes_CI.apply(lambda res: res.high).fillna(\"NaN\")\n",
    "\n",
    "\n",
    "#  Double CN counts group - interval - <low, high>\n",
    "\n",
    "main_genes_copy[\"OR2_low\"] = OR2_genes_CI.apply(lambda res: res.low).fillna(\"NaN\")\n",
    "main_genes_copy[\"OR2_high\"] = OR2_genes_CI.apply(lambda res: res.high).fillna(\"NaN\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Formatting and export final output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Final files - stats_exon; main_genes_copy\n",
    "\n",
    "# necessary columns per \n",
    "\n",
    "# poziční - CHROM, POS, REF, ALT, ID\n",
    "# FORMAT - data experimentálních subjektů | upřímě nevím..\n",
    "# Fst - fst statitika \n",
    "# OR - odds ratio | dle počtu CN\n",
    "# Fisher exact test - FET  | dle počtu CN\n",
    "\n",
    "\n",
    "# CN - regular\n",
    "\n",
    "col_stats_1 = [\"CHROM\",\"POS\",\"REF\",\"ALT\",\"ID\",\"GENE\",\"HGVS_c\",\"HGVS_p\",\"EFFECT\",\"IMPACT\",\"CS_AA\",\"CS_AR\",\"CS_RR\",\"CN_AA\",\"CN_AR\", \"CN_RR\",\"CS_A\",\"CS_R\",\"CN_A\",\"CN_R\",\"CN_2A\",\"CN_2R\", \"OR1\",\"FET1\",\"OR1_low\",\"OR1_high\",\"Fst\"]\n",
    "\n",
    "# CN - double\n",
    "\n",
    "col_stats_2 = [\"CHROM\",\"POS\",\"REF\",\"ALT\",\"ID\",\"GENE\",\"HGVS_c\", \"HGVS_p\",\"EFFECT\",\"IMPACT\",\"CS_AA\",\"CS_AR\",\"CS_RR\",\"CN_AA\",\"CN_AR\", \"CN_RR\",\"CS_A\",\"CS_R\",\"CN_A\",\"CN_R\",\"CN_2A\",\"CN_2R\",\"FET1\",\"FET2\",\"OR1\",\"OR1_low\",\"OR1_high\",\"OR2\",\"OR2_low\",\"OR2_high\",\"Fst\"]\n",
    "\n",
    "# main_exones_copy\n",
    "\n",
    "\n",
    "# exones - FET1\n",
    "\n",
    "stats_exones_FET1 = main_exones_copy.loc[:,col_stats_1]\n",
    "\n",
    "# exones - FET2\n",
    "\n",
    "stats_exones_FET2 = main_exones_copy.loc[:,col_stats_2]\n",
    "\n",
    "\n",
    "# main_genes_copy\n",
    "\n",
    "\n",
    "# genes - FET1\n",
    "\n",
    "stats_genes_FET1 = main_genes_copy.loc[:,col_stats_1]\n",
    "\n",
    "# genes - FET2\n",
    "\n",
    "stats_genes_FET2 = main_genes_copy.loc[:,col_stats_2]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Final statistical output |\n",
    "\n",
    "# stats_exones_FET\n",
    "\n",
    "stats_exones_FET1.to_excel(\"StatsExones_FET1_single.xlsx\", index=False)\n",
    "stats_exones_FET2.to_excel(\"StatsExones_FET2_single.xlsx\", index=False)\n",
    "\n",
    "# stats_genes_FET\n",
    "\n",
    "stats_genes_FET1.to_excel(\"StatsGenes_FET1_single.xlsx\", index=False)\n",
    "stats_genes_FET2.to_excel(\"StatsGenes_FET2_single.xlsx\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
